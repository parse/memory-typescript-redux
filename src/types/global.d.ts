declare module 'enzyme-adapter-react-16';
declare module 'redux-saga-test-plan';
