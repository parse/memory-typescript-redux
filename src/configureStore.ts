import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension';
import { StoreState } from './types/index';
import gameSaga from './sagas/game';
import notificationSaga from './sagas/notification';
import rootReducer from './reducers';

export default function configureStore(initialState?: StoreState) {
  const sagaMiddleware = createSagaMiddleware();

  const internalStore = createStore(
    rootReducer,
    initialState,
    composeWithDevTools(applyMiddleware(sagaMiddleware))
  );

  sagaMiddleware.run(gameSaga);
  sagaMiddleware.run(notificationSaga);

  return internalStore;
}
