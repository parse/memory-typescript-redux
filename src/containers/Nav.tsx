import * as React from 'react';
import * as actions from '../actions';

import { connect, Dispatch } from 'react-redux';
import { bindActionCreators } from 'redux';

import { StoreState, Tile as TileType, GameStatus } from '../types/index';

interface NavProps {
  startGame: () => void;
  numberOfTries: number;
  status: GameStatus;
}

class Nav extends React.Component<NavProps, {}> {
  componentDidMount() {
    const { startGame } = this.props;

    startGame();
  }

  handleClickResetButton() {
    const { startGame } = this.props;

    setTimeout(() => {
      startGame();
    }, 1000);
  }

  renderGameStatus() {
    const { status } = this.props;

    return (
      <p className="navbar-brand">
        {status === GameStatus.Ongoing && <p>Game is ongoing</p>}
        {status === GameStatus.Ended && <p>Game has ended</p>}
      </p>
    );
  }

  render() {
    const { numberOfTries } = this.props;

    return (
      <nav className="navbar navbar-default navbar-top">
        <div className="container">
          <div className="navbar-header">
            <p className="navbar-brand">
              Number of tries:{' '}
              <span className="label label-success">{numberOfTries}</span>
            </p>
            {this.renderGameStatus()}
          </div>
          <div className="collapse navbar-collapse">
            <ul className="nav navbar-nav navbar-right">
              <li>
                <button onClick={() => this.handleClickResetButton()}>
                  Restart
                </button>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    );
  }
}

function mapStateToProps({ numberOfTries, status }: StoreState) {
  return {
    numberOfTries,
    status,
  };
}

function mapDispatchToProps(dispatch: Dispatch<actions.MemoryAction>) {
  return bindActionCreators(
    {
      startGame: actions.startGame,
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(Nav as any);
