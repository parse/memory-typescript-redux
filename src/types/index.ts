export enum GameStatus {
  NotStarted = 0,
  Ongoing = 1,
  Ended = 2,
}

export enum Severity {
  Info,
  Success,
  Error,
}

export interface StoreState {
  readonly memory: MemoryState;
  readonly notification: NotificationState;
}

export interface MemoryState {
  readonly tiles: Array<Tile>;
  readonly isWaiting: Boolean;
  readonly numberOfTries: number;
  readonly status: GameStatus;
}

export interface NotificationState {
  readonly notifications: Array<Notification>;
}

export interface Tile {
  readonly image: String;
  readonly tileKey: number;
  readonly flipped: Boolean;
  readonly matched: Boolean;
}

export interface Notification {
  readonly uid: number;
  readonly message: String;
  readonly severity: Severity;
}
