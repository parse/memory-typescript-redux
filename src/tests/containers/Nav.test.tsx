import * as React from 'react';
import { Provider } from 'react-redux';
import { GameStatus } from '../../types/index';

import Nav from '../../containers/Nav';
import * as renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import configureStore from '../../store/';

describe('Nav', () => {
  let store;
  let tree;

  it('renders correctly game status ENDED', () => {
    const state = {
      memory: {
        tiles: [],
        isWaiting: false,
        numberOfTries: 0,
        status: GameStatus.Ended,
      },
      notification: { notifications: [] },
    };

    store = configureStore(state);
    tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Nav />
          </BrowserRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly game status ONGOING', () => {
    const state = {
      memory: {
        tiles: [],
        isWaiting: false,
        numberOfTries: 0,
        status: GameStatus.Ongoing,
      },
      notification: { notifications: [] },
    };

    store = configureStore(state);
    tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Nav />
          </BrowserRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
