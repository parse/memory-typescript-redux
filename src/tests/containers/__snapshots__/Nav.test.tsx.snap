// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav renders correctly game status ENDED 1`] = `
<nav
  className="navbar navbar-default navbar-top"
>
  <div
    className="container"
  >
    <div
      className="navbar-header"
    >
      <p
        className="navbar-brand"
      >
        Number of tries:
         
        <span
          className="label label-success"
        >
          0
        </span>
      </p>
      <p
        className="navbar-brand"
      >
        <span
          className="label label-info"
        >
          Game is ongoing
        </span>
      </p>
    </div>
    <div
      className="collapse navbar-collapse"
    >
      <ul
        className="nav navbar-nav navbar-right"
      >
        <li>
          <a
            aria-current="true"
            className="active"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            Game
          </a>
        </li>
        <li>
          <a
            aria-current={false}
            className={undefined}
            href="/about"
            onClick={[Function]}
            style={undefined}
          >
            About
          </a>
        </li>
        <li>
          <p
            className="navbar-btn"
          >
            <a
              className="btn btn-primary"
              onClick={[Function]}
            >
              Restart
            </a>
          </p>
        </li>
      </ul>
    </div>
  </div>
</nav>
`;

exports[`Nav renders correctly game status ONGOING 1`] = `
<nav
  className="navbar navbar-default navbar-top"
>
  <div
    className="container"
  >
    <div
      className="navbar-header"
    >
      <p
        className="navbar-brand"
      >
        Number of tries:
         
        <span
          className="label label-success"
        >
          0
        </span>
      </p>
      <p
        className="navbar-brand"
      >
        <span
          className="label label-info"
        >
          Game is ongoing
        </span>
      </p>
    </div>
    <div
      className="collapse navbar-collapse"
    >
      <ul
        className="nav navbar-nav navbar-right"
      >
        <li>
          <a
            aria-current="true"
            className="active"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            Game
          </a>
        </li>
        <li>
          <a
            aria-current={false}
            className={undefined}
            href="/about"
            onClick={[Function]}
            style={undefined}
          >
            About
          </a>
        </li>
        <li>
          <p
            className="navbar-btn"
          >
            <a
              className="btn btn-primary"
              onClick={[Function]}
            >
              Restart
            </a>
          </p>
        </li>
      </ul>
    </div>
  </div>
</nav>
`;
